feat: add `blas/ext/last-index-of`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Progreses #2656.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `blas/ext/last-index-of`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add blas/ext/last-index-of

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+// returns <ndarray>
+
+// Perform operation:
+var out = lastIndexOf( x, 2.0, 3 );
+// returns <ndarray>
+
+var idx = out.get();
+// returns 4
```
```suggestion
var out = lastIndexOf( x, 1.0, 3 );
// returns <ndarray>

var idx = out.get();
// returns -1
```

headlessNode committed:
docs: refactor the example

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
@headlessNode You'll want to make similar changes as done in https://github.com/stdlib-js/stdlib/pull/7565 in this PR.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte done

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1655 @@
+	t.end();
+});
+
+tape( 'the function supports providing a from index (scalar, broadcasted)', function test( t ) {
```
I don't think this is the test that you want. You have already tested "broadcasting" of a scalar from index above. Here, you are testing that broadcasting works when specifying a dimension. That is fine, but also does not cover the case where `fromIndex` is an ndarray which is NOT broadcasted. Here, that would be `fromIndex` is a two-element vector, which would test the case where we use a different from index value for the two columns.

You can choose to keep this and the following tests, but you need to update the descriptions, and, at a minimum, you need to add tests for when the from index argument is not broadcasted and using different from indices. You should also test similarly for the search element (i.e., meaning a test where the search element is different for each searched row/column).

kgryte replied:
This same comment may be applicable for `blas/ext/index-of`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,223 @@
+	};
+
+	// Initialize the `fromIndex` to the first element along a dimension:
+	fidx = 0;
```
I think we've initialized the `fromIndex` incorrectly here, as this doesn't follow https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/lastIndexOf semantics. There, `fromIndex` specifies the first element at which to begin searching backward. Here, `fromIndex` is used similar to `indexOf` and something like a means for reducing the sub-array.

This also needs to be addressed in `blas/ext/base/*last-index-of`, as, looking at examples, those all don't appear to be parameterized correctly based upon conventional semantics.

kgryte reviewed (changes requested):
Left an initial review.

headlessNode committed:
Merge branch 'stdlib-js:develop' into top-last-index-of

headlessNode committed:
fix: test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: apply suggestions form code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add missing test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1758 @@
+	}
+});
+
+tape( 'the function throws an error if provided a third argument which is not an ndarray-like object or an integer', function test( t ) {
```
```suggestion
tape( 'the function throws an error if provided an output argument which is not an ndarray-like object or an integer', function test( t ) {
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1758 @@
+	x = zeros( [ 2, 2 ], {
+		'dtype': 'float64'
+	});
+	y = zeros( [], {
+		'dtype': 'generic'
+	});
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1758 @@
+	var values;
+	var i;
+	var x;
+	var y;
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1758 @@
+
+	function badValue( value ) {
+		return function badValue() {
+			lastIndexOf( x, scalar2ndarray( 2.0 ), value, y, {} );
```
```suggestion
			lastIndexOf( x, scalar2ndarray( 2.0 ), scalar2ndarray( 0 ), value, {} );
```

headlessNode committed:
fix: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+// returns <ndarray>
+
+var idx = out.get();
+// returns -1
```
@headlessNode This doesn't seem correct. Shouldn't the answer be `3`?

kgryte replied:
```suggestion
// returns 3
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+
+-   **x**: input [ndarray][@stdlib/ndarray/ctor]. Must have at least one dimension.
+-   **searchElement**: search element. May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor]. If provided a scalar value, the value is cast to the data type of the input [ndarray][@stdlib/ndarray/ctor]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the non-reduced dimensions of t ...
+-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an integer index or "generic" [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the non-reduced dimensions of the inp ...
```
```suggestion
-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an integer index or "generic" [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the non-reduced dimensions of the input [ndarray][@stdlib/ndarray/ctor]. For example, given the input shape `[2, 3, 4]` and `options.dim=0`, a provided [ndarray][@stdlib/ndarray/ctor] must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the shape `[3, 4]`. If provided a negative integer, the index at which to begin searching along a dimension is determined by counting backward from the last element (where `-1` refers to the last element). Default: `-1`.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
````diff
@@ -0,0 +1,260 @@
+// returns -1
+```
+
+By default, the function begins searching from the first element along the reduction dimension. To begin searching from a different index, provide a `fromIndex` argument.
````
```suggestion
By default, the function begins searching from the last element along the reduction dimension. To begin searching from a different index, provide a `fromIndex` argument.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+// returns <ndarray>
+
+var idx = out.get();
+// returns 4
```
```suggestion
var out = lastIndexOf( x, 2.0, 3 );
// returns <ndarray>

var idx = out.get();
// returns 2
```

Your current example just replicates the default behavior.

kgryte committed:
docs: fix return value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix default

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+// returns <ndarray>
+
+var idx = ndarray2array( out );
+// returns [ 0, -1 ]
```
```suggestion
// returns [ 1, -1 ]
```

Shouldn't this be `[ 1, -1]`, if the matrix is

```
[
  [ -3.0, 2.0 ],
  [ -3.0, 4.0 ]
]
```

where `dim=0`, this means down the columns. The last matching element is element `x(1,0)`, not `x(0,0)`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+// returns <ndarray>
+
+var idx = ndarray2array( out );
+// returns [ [ 0, -1 ] ]
```
Same thing.

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+// returns <ndarray>
+
+var idx = out.get();
+// returns -1
```
```suggestion
// returns 3
```

kgryte committed:
docs: fix return value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/README.md:
```diff
@@ -0,0 +1,260 @@
+
+-   **x**: input [ndarray][@stdlib/ndarray/ctor]. Must have at least one dimension.
+-   **searchElement**: search element. May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor]. If provided a scalar value, the value is cast to the data type of the input [ndarray][@stdlib/ndarray/ctor]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the non-reduced dimensions of t ...
+-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an integer index or "generic" [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the non-reduced dimensions of the inp ...
```
```suggestion
-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an integer index or "generic" [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the non-reduced dimensions of the input [ndarray][@stdlib/ndarray/ctor]. For example, given the input shape `[2, 3, 4]` and `options.dim=0`, a provided [ndarray][@stdlib/ndarray/ctor] must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the shape `[3, 4]`. If provided a negative integer, the index at which to begin searching along a dimension is determined by counting backward from the last element (where `-1` refers to the last element). Default: `-1`.
```

kgryte committed:
docs: fix default

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented:
/stdlib merge

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/base.js:
```diff
@@ -0,0 +1,120 @@
+* // returns <ndarray>
+*
+* var idx = out.get();
+* // returns -1
```
```suggestion
* // returns 3
```

@headlessNode I am not sure why everything returning `-1` wasn't a red flag for you. There seems to be something wrong with the current implementation.

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into top-last-index-of

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,202 @@
+	};
+
+	// Initialize the `fromIndex` to the first element along a dimension:
+	fidx = 0;
```
```suggestion
	// Initialize the `fromIndex` to the first element along a dimension:
	fidx = -1;
```

kgryte committed:
fix: update default value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,203 @@
+* // returns <ndarray>
+*
+* var arr = ndarray2array( out );
+* // returns [ -1, 0 ]
```
```suggestion
* // returns [ -1, 0 ]
```

This isn't correct either.

```
[
  1.0, 2.0, -3.0,
  2.0, -5.0, 6.0
]
```

Over the last dimension means across the columns. In which case, I'd expect `[ 1, 0 ]`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,203 @@
+	};
+
+	// Initialize the `fromIndex` to the first element along a dimension:
+	// Initialize the `fromIndex` to the first element along a dimension:
```
```suggestion
```

kgryte committed:
docs: remove duplicate comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,202 @@
+		'keepdims': false
+	};
+
+	// Initialize the `fromIndex` to the first element along a dimension:
```
```suggestion
	// Initialize the `fromIndex` to the last element along a dimension:
```

kgryte committed:
docs: fix comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,223 @@
+	};
+
+	// Initialize the `fromIndex` to the first element along a dimension:
+	fidx = 0;
```
```suggestion
	// Initialize the `fromIndex` to the last element along a dimension:
	fidx = -1;
```

kgryte committed:
fix: update default value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/base.js:
```diff
@@ -0,0 +1,120 @@
+* })
+*
+* // Create a from index ndarray:
+* var fromIndex = scalar2ndarray( 0, {
```
```suggestion
* var fromIndex = scalar2ndarray( -1, {
```

kgryte committed:
docs: fix example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix return value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
@headlessNode You need to go back and do a closer review of this PR, as it still contains a number of mistakes.

headlessNode committed:
fix: refactor docs & tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: lint error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte unexpected `no-unnecessary-nested-functions` lint error.

kgryte commented:
@headlessNode Yes, a new lint rule was added that is buggy. You can ignore for now.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into top-last-index-of

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/docs/repl.txt:
```diff
@@ -0,0 +1,132 @@
+        shape which is broadcast-compatible with the shape `[3, 4]`. If provided
+        a negative integer, the index at which to begin searching along a
+        dimension is determined by counting backward from the last element
+        (where -1 refers to the last element). Default: 0.
```
```suggestion
        (where -1 refers to the last element). Default: -1.
```

kgryte committed:
docs: fix default

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/docs/repl.txt:
```diff
@@ -0,0 +1,132 @@
+        shape which is broadcast-compatible with the shape `[3, 4]`. If provided
+        a negative integer, the index at which to begin searching along a
+        dimension is determined by counting backward from the last element
+        (where -1 refers to the last element). Default: 0.
```
```suggestion
        (where -1 refers to the last element). Default: -1.
```

kgryte committed:
docs: fix default

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,223 @@
+*
+* @param {ndarrayLike} x - input ndarray
+* @param {(ndarrayLike|*)} searchElement - search element
+* @param {(ndarrayLike|integer)} [fromIndex] - index from which to begin searching
```
```suggestion
* @param {(ndarrayLike|integer)} [fromIndex=-1] - index from which to begin searching
```

kgryte committed:
docs: add default

Signed-off-by: Athan <kgryte@gmail.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,202 @@
+* @param {(ndarrayLike|*)} searchElement - search element
+* @param {(ndarrayLike|integer)} [fromIndex] - index from which to begin searching
+* @param {Options} [options] - function options
+* @param {IntegerArray} [options.dim=-1] - dimension over which to perform operation
```
```suggestion
* @param {Integer} [options.dim=-1] - dimension over which to perform operation
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,223 @@
+
+	nargs = arguments.length;
+	if ( !isndarrayLike( x ) ) {
+		throw new TypeError( format( 'invalid argument. The first argument must be an ndarray. Value: `%s`.', x ) );
```
```suggestion
		throw new TypeError( format( 'invalid argument. First argument must be an ndarray. Value: `%s`.', x ) );
```

kgryte committed:
fix: update error message

Signed-off-by: Athan <kgryte@gmail.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,223 @@
+* @param {(ndarrayLike|integer)} [fromIndex=-1] - index from which to begin searching
+* @param {ndarrayLike} out - output ndarray
+* @param {Options} [options] - function options
+* @param {IntegerArray} [options.dim=-1] - dimension over which to perform operation
```
```suggestion
* @param {Integer} [options.dim=-1] - dimension over which to perform operation
```

headlessNode committed:
fix: parameter description

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,223 @@
+	// Resolve provided options...
+	if ( hasOwnProp( options, 'dim' ) ) {
+		opts.dims[ 0 ] = options.dim;
+	}
```
```suggestion
	if ( hasOptions ) {
		if ( !isPlainObject( options ) ) {
			throw new TypeError( format( 'invalid argument. Options argument must be an object. Value: `%s`.', options ) );
		}
		// Resolve provided options...
		if ( hasOwnProp( options, 'dim' ) ) {
			opts.dims[ 0 ] = options.dim;
		}
	}
```

headlessNode committed:
fix: parameter description

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte committed:
refactor: avoid unnecessarily calling `hasOwnProp`

Signed-off-by: Athan <kgryte@gmail.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/assign.js:
```diff
@@ -0,0 +1,225 @@
+* @param {(ndarrayLike|integer)} [fromIndex=-1] - index from which to begin searching
+* @param {ndarrayLike} out - output ndarray
+* @param {Options} [options] - function options
+* @param {Integer} [options.dim=-1] - dimension over which to perform operation
```
```suggestion
* @param {integer} [options.dim=-1] - dimension over which to perform operation
```

headlessNode committed:
fix: parameter type description

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,202 @@
+* @param {(ndarrayLike|*)} searchElement - search element
+* @param {(ndarrayLike|integer)} [fromIndex] - index from which to begin searching
+* @param {Options} [options] - function options
+* @param {Integer} [options.dim=-1] - dimension over which to perform operation
```
```suggestion
* @param {integer} [options.dim=-1] - dimension over which to perform operation
```

headlessNode committed:
fix: parameter type description

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,202 @@
+*
+* @param {ndarrayLike} x - input ndarray
+* @param {(ndarrayLike|*)} searchElement - search element
+* @param {(ndarrayLike|integer)} [fromIndex] - index from which to begin searching
```
```suggestion
* @param {(ndarrayLike|integer)} [fromIndex=-1] - index from which to begin searching
```

kgryte committed:
docs: document default value

Signed-off-by: Athan <kgryte@gmail.com>

headlessNode commented:
🫣

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/lib/main.js:
```diff
@@ -0,0 +1,202 @@
+	}
+	if ( hasOwnProp( options, 'dtype' ) ) {
+		opts.dtype = options.dtype;
+	}
```
```suggestion
	if ( hasOptions ) {
		if ( !isPlainObject( options ) ) {
			throw new TypeError( format( 'invalid argument. Options argument must be an object. Value: `%s`.', options ) );
		}
		// Resolve provided options...
		if ( hasOwnProp( options, 'dim' ) ) {
			opts.dims[ 0 ] = options.dim;
		}
		if ( hasOwnProp( options, 'keepdims' ) ) {
			opts.keepdims = options.keepdims;
		}
		if ( hasOwnProp( options, 'dtype' ) ) {
			opts.dtype = options.dtype;
		}
	}
```

kgryte committed:
fix: avoid repeatedly calling `hasOwnProp` when we don't need to

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1754 @@
+	}
+});
+
+tape( 'the function throws an error if provided an output argument which is not an ndarray-like object or an integer', function test( t ) {
```
This description doesn't make sense. An output argument cannot be an integer.

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1754 @@
+	}
+});
+
+tape( 'the function throws an error if provided an output argument which is not an ndarray-like object or an integer', function test( t ) {
+	var values;
+	var i;
+	var x;
+
+	x = zeros( [ 2, 2 ], {
+		'dtype': 'float64'
+	});
+	values = [
+		'5',
+		NaN,
+		true,
+		false,
+		null,
+		void 0,
+		[],
+		function noop() {}
+	];
+	for ( i = 0; i < values.length; i++ ) {
+		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			lastIndexOf( x, scalar2ndarray( 2.0 ), scalar2ndarray( 0 ), value, {} );
+		};
+	}
+});
```
```suggestion
tape( 'the function throws an error if provided an output argument which is not an ndarray-like object', function test( t ) {
	var values;
	var opts;
	var i;
	var x;

	opts = {
		'dtype': 'generic'
	};

	x = zeros( [ 2, 2 ], opts );

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			lastIndexOf( x, 2.0, value );
		};
	}
});
```

kgryte committed:
test: fix test

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1760 @@
+	}
+});
+
+tape( 'the function throws an error if provided an output argument which is not an ndarray-like object', function test( t ) {
+	var values;
+	var opts;
+	var i;
+	var x;
+
+	opts = {
+		'dtype': 'generic'
+	};
+
+	x = zeros( [ 2, 2 ], opts );
+
+	values = [
+		'5',
+		5,
+		NaN,
+		true,
+		false,
+		null,
+		void 0,
+		[],
+		{},
+		function noop() {}
+	];
+	for ( i = 0; i < values.length; i++ ) {
+		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			lastIndexOf( x, 2.0, value );
+		};
+	}
+});
+
```
```suggestion
```

kgryte committed:
test: remove dupe test

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: fix descriptions and clean-up

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.main.js:
```diff
@@ -0,0 +1,1458 @@
+	var x;
+
+	xbuf = [ -1.0, 2.0, -3.0, 2.0 ];
+	x = new ndarray( 'generic', xbuf, [ 2, 2 ], [ 2, 1 ], 0, 'column-major' );
```
```suggestion
	x = new ndarray( 'generic', xbuf, [ 2, 2 ], [ 1, 2 ], 0, 'column-major' );
```

kgryte committed:
test: fix strides

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.main.js:
```diff
@@ -0,0 +1,1458 @@
+	actual = lastIndexOf( x, 2.0, {
+		'keepdims': true
+	});
+	expected = [ [ 1 ], [ 1 ] ];
```
```suggestion
	expected = [ [ -1 ], [ 1 ] ];
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/last-index-of/test/test.main.js:
```diff
@@ -0,0 +1,1458 @@
+	actual = lastIndexOf( x, 2.0, 0, {
+		'keepdims': true
+	});
+	expected = [ [ -1 ], [ -1 ] ];
```
```suggestion
	expected = [ [ -1 ], [ 0 ] ];
```

kgryte committed:
test: fix expected value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: fix expected value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: update test descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up. I also went ahead and incorporated some of the test changes in `index-of`.
